preprocessing/nextclade/src/loculus_preprocessing/embl.py

```diff
@@ -0,0 +1,226 @@
+            cds_qualifiers = getattr(config, "annotations", {}).get("cds_qualifiers", [])
+            cds_attributes_map = {qualifier: qualifier for qualifier in cds_qualifiers}
+            cds_attributes_map.update(attribute_map)
+            segments = cds.get("segments", [])
+            ranges = [segment.get("range") for segment in segments]
+            attributes_cds = cds.get("attributes", {})
+            strands = [-1 if segment.get("strand") == "-" else +1 for segment in segments]
+            locations = [
+                FeatureLocation(start=r["begin"], end=r["end"], strand=s)
+                for r, s in zip(ranges, strands, strict=False)
+            ]
+            compound_location = locations[0] if len(locations) == 1 else locations
+            qualifiers = {
+                new_key: attributes_cds[old_key]
+                for old_key, new_key in cds_attributes_map.items()
+                if old_key in attributes_cds
+            }
+            qualifiers["codon_start"] = 1
+            qualifiers["translation"] = "".join(
+                [
+                    str(Seq(sequence_str[(range["begin"]) : (range["end"])]).translate())
+                    for range in ranges
+                ]
+            )
+            feature = SeqFeature(
+                location=None,  # TODO - compound_location caused a type error
+                # location=compound_location,
+                type="CDS",
+                qualifiers=qualifiers,
+            )
+            feature_list.append(feature)
+    return feature_list
+
+
+def create_flatfile(
+    config,
+    accession,
+    version,
+    metadata,
+    organism_metadata,
+    unaligned_nucleotide_sequences,
+    dir,
+    annotation_object: dict[str, Any] | None = None,
+):
+    collection_date = metadata.get("sampleCollectionDate", "Unknown")
+    country = get_country(metadata)
+    organism = organism_metadata.get("scientific_name", "Unknown")
+    description = get_description(accession, version, config.db_name)
+    authors = get_authors(metadata.get("authors", ""))
+    moleculetype = get_molecule_type(organism_metadata)
+
+    if dir:
+        os.makedirs(dir, exist_ok=True)
+        filename = os.path.join(dir, "sequences.embl")
+    else:
+        with tempfile.NamedTemporaryFile(delete=False, suffix=".embl") as temp:
+            filename = temp.name
+
+    seqIO_moleculetype = {  # noqa: N806
+        "GENOMIC_DNA": "DNA",
+        "GENOMIC_RNA": "RNA",
+        "VIRAL_CRNA": "cRNA",
+    }
+
+    embl_content = []
+
+    multi_segment = set(unaligned_nucleotide_sequences.keys()) != {"main"}
+
+    for seq_name, sequence_str in unaligned_nucleotide_sequences.items():
+        if not sequence_str:
+            continue
+        reference = Reference()
+        reference.authors = authors
+        sequence = SeqRecord(
+            Seq(sequence_str),
+            id=f"{accession}_{seq_name}" if multi_segment else accession,
+            annotations={},  # TODO
```

Comment:
btw pylance throws a type error for the annotations dict but you can ignore it

Reply:
ah good to know, thanks!